Add evergreen blog: Kubernetes Configuration Best Practices (#53172)
kubernetes/website: content/en/blog/_posts/2025-11-11-kubernetes-configuration-best-practices/index.md

---
layout: blog
title: "Kubernetes Configuration Best Practices"

> date: 2025-11-11

Reviewer: ```suggestion
date: 2025-11-12T09:00:00+05:00
```

slug: kubernetes-configuration-best-practices
evergreen: true
author: Kirti Goyal

> draft: false

Reviewer: ```suggestion
draft: true
```

We can merge a follow-up PR that adds the redirect and removes the original page.

Avoid setting default values that are already handled by Kubernetes. Minimal YAMLs are easier to debug, cleaner to review and less likely to break things later. 

###	Group related objects together
If your Deployment, Service and ConfigMap all belong to one app, put them in a single YAML file.  It's easier to track changes and apply them as a unit. See the [Guestbook all-in-one.yaml](https://github.com/kubernetes/examples/blob/master/web/guestbook/all-in-one/guestbook-all-in-one.yaml) file as an example of this syntax.

You can even apply entire directories with:
```bash
kubectl apply -f configs/
```
One command and boom everything in that folder gets deployed. 

###	Add helpful annotations
YAMl files are not just for machines, they are for humans too. Use annotations to describe why something exists or what it does. A quick one-liner can save hours when debugging later and also allows better collaboration.  

## Managing Workloads: Pods, Deployments, and Jobs

Most of us started our Kubernetes journey by creating Pods directly. 
While Pods are the building blocks of Kubernetes, using them directly is like driving without a seatbelt, it works until something crashes.

> "Naked Pods" (Pods not managed by a controller, [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) or a [ReplicaSet](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/)) are fine for testing, but in real setups, they are risky. 

Reviewer: ```suggestion
_Naked Pods_ (Pods not managed by a controller, such as [Deployment](/docs/concepts/workloads/controllers/deployment/) or a [StatefulSet](/docs/concepts/workloads/controllers/statefulset/)) are fine for testing, but in real setups, they are risky.

```

Why?
Because if the node hosting that Pod dies, the Pod dies with it and Kubernetes won't bring it back automatically. 

### Use Deployments for apps that should always be running

> A Deployment, which both creates a ReplicaSet to ensure that the desired number of Pods is always available, and specifies a strategy to replace Pods (such as [RollingUpdate](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-update-deployment)), is almost always preferable to creating Pods directly. 

Reviewer: ```suggestion
A Deployment, which both creates a ReplicaSet to ensure that the desired number of Pods is always available, and specifies a strategy to replace Pods (such as [RollingUpdate](/docs/concepts/workloads/controllers/deployment/#rolling-update-deployment)), is almost always preferable to creating Pods directly.
```

You can roll out a new version, and if something breaks, roll back instantly.

### Use Jobs for tasks that should finish
A [Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) is perfect when you need something to run once and then stop like database migration or batch processing task. 
It will retry if the pods fails and report success when it's done. 

## Service Configuration and Networking

Services are how your workloads talk to each other inside (and sometimes outside) your cluster. Without them, your pods exist but can't reach anyone. Let's make sure that doesn't happen

### Create Services before workloads that use them
When Kubernetes starts a Pod, it automatically injects environment variables for existing Services.

>  So, if a Pod depends on a Service. Create a [Service](https://kubernetes.io/docs/concepts/services-networking/service/) before its corresponding backend workloads (Deployments or ReplicaSets), and before any workloads that need to access it.  For example, if a Service named foo exists, all containers will get the following variables in their initial environment:

Reviewer: ```suggestion
So, if a Pod depends on a Service, create a [Service](/docs/concepts/services-networking/service/) **before** its corresponding backend workloads (Deployments or StatefulSets), and before any workloads that need to access it.

For example, if a Service named foo exists, all containers will get the following variables in their initial environment:
```

Unless you're debugging or building something like a network plugin, avoid them. 
If you just need local access for testing, try [`kubectl port-forward`](https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/) :
```bash
kubectl port-forward deployment/web 8080:80
```
Or if you really need external access, use a [NodePort Service](https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport). That's the safer, Kubernetes-native way. 

### Use headless Services for internal discovery 
Sometimes, you don't want Kubernetes to load balance traffic. You want to talk directly to each Pod. That's where [headless Services](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) come in. 

You create one by setting `clusterIP: None`.
Instead of a single IP, DNS gives you a list of all Pods IPs, perfect for apps that manage connections themselves. 

> ## Working with Labels Effectively 

Reviewer: ```suggestion
## Working with labels effectively
```

> Labels are key/value pairs that are attached to objects such as Pods. 

Reviewer: ```suggestion
[Labels](/docs/concepts/overview/working-with-objects/labels/) are key/value pairs that are attached to objects such as Pods.
```

###	Manipulate labels for debugging 
Since controllers (like ReplicaSets or Deployments) use labels to manage Pods, you can remove a label to “detach” a Pod temporarily.

Example:
```bash
kubectl label pod mypod app-
```
The `app-` part removes the label key `app`.
Once that happens, the controller won’t manage that Pod anymore.
It’s like isolating it for inspection, a “quarantine mode” for debugging. To interactively remove or add labels, use [`kubectl label`](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#label).

You can then check logs, exec into it and once done, delete it manually.

That’s a super underrated trick every Kubernetes engineer should know.

## Handy kubectl Tips for Managing Configs

These small tips make life much easier when you are working with multiple YAMLs or clusters.

### Apply entire directories
Instead of applying one file at a time, apply the whole folder:
```bash
kubectl apply -f configs/
```
This command looks for `.yaml`, `.yml` and `.json` files in that folder and applies them all together.
It's faster, cleaner and helps keep things grouped by app. 

### Use label selectors to get or delete resources

> You don't always need to type object names. Instead, use [selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors) to act on entire groups at once:

Reviewer: ```suggestion
You don't always need to type out resource names one by one.
Instead, use [selectors]/docs/concepts/overview/working-with-objects/labels/#label-selectors) to act on entire groups at once:
```

```bash
kubectl get pods -l app=myapp
kubectl delete pod -l phase=test
```
It's especially useful in CI/CD pipelines, where you want to clean up test resources dynamically. 

### Quickly create deployments and services
For quick experiments, you don't always need to write YAMLs. You can spin up a deployment right from the CLI:

```bash
kubectl create deployment webapp --image=nginx
```

Then expose it as a Service:
```bash
kubectl expose deployment webapp --port=80
```
This is great when you just want to test something before writing full manifests. 

> Also, see [Use a Service to Access an Application in a cluster](https://kubernetes.io/docs/tasks/access-application-cluster/service-access-application-cluster/) for an example. 

Reviewer: ```suggestion
Also, see [Use a Service to Access an Application in a cluster](/docs/tasks/access-application-cluster/service-access-application-cluster/) for an example.
```

## Conclusion

> Clean configuration leads to calm clusters. If you stick to a few simple habits: version-control everything, use consistent labels, prefer YAML over JSON and avoid naked Pods you'll save yourself hours of debugging down the road.

Reviewer: (optional)
```suggestion
Cleaner configuration leads to calmer cluster administrators. 
If you stick to a few simple habits: keep configuration simple and minimal, version-control everything,
use consistent labels, and avoid relying on naked Pods, you'll save yourself hours of debugging down the road.
```

The best part? 

> Good configurations age well. Even months later, you (or others) will be able to read your manifests and understand exactly what's happening without confusion or chaos. 

Reviewer: This reads a lot like a robot wrote it. Can you reword it?

Whether you are just starting out or already deploying apps daily, these are the little things that keep your cluster stable and your future self sane. 

## General Configuration Practices

### Use the latest stable API version 
Kubernetes evolves fast. Older APIs eventually get deprecated and stop working. So, whenever you are defining resources, make sure you are using the latest stable API version. 
You can always check with
```bash
kubectl api-resources
```
This simple step saves you from future compatibility issues. 
  
### Store configuration in version control 
Never apply YAML files directly from your desktop. Always keep them in a version control system like Git, it's your safety net. 
If something breaks, you can instantly roll back to a previous commit, compare changes or recreate your cluster setup without panic.

### Write configs in YAML not JSON
Write your configuration files using YAML rather than JSON. Both work technically, but YAML is just easier for humans. It's cleaner to read and less noisy and widely used in the community. 

> Note: YAML has some sneaky gotchas with booleans. 
Use only `true` or `false`. 
Don't write `yes`, `no`, `on` or  `off`.

> They might work in one version of YAML but break in another. To be safe, quote anything that looks like a Boolean (for example `"yes"`).

Author: got it

---
layout: blog
title: "Kubernetes Configuration Good Practices"
date: 2025-11-11
slug: kubernetes-configuration-good-practices
evergreen: true

> author: Kirti Goyal

Reviewer: We may want to acknowledge other contributors' work on the original page.
